File: profiles/java-spring/scaffold/src/main/java/com/example/app/NotFoundAdvice.java
package com.example.app;

import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.servlet.NoHandlerFoundException;

/**
 * Turns an unmatched GET route into the hardened, framework-neutral 404 the parity contract fixes.
 *
 * <p>With {@code spring.mvc.throw-exception-if-no-handler-found=true} and {@code
 * spring.web.resources.add-mappings=false} an unknown path raises {@link NoHandlerFoundException}
 * rather than serving the default whitelabel error page; this advice maps it to a compact {@code
 * {"error":"not found"}} JSON body — no stack, no framework/version leak. (Non-GET methods never
 * reach the dispatcher: {@link AppInstrumentationFilter} short-circuits them to the same 404.)
 */
@RestControllerAdvice
public class NotFoundAdvice {

  /**
   * The compact 404 payload — matches the go/python spine byte-for-byte after Jackson renders it.
   */
  private static final Map<String, String> NOT_FOUND_BODY = Map.of("error", "not found");

  /**
   * Maps an unmatched route to the hardened JSON 404.
   *
   * @return a 404 carrying {@code {"error":"not found"}}
   */
  @ExceptionHandler(NoHandlerFoundException.class)
  public ResponseEntity<Map<String, String>> handleNotFound() {
    return ResponseEntity.status(HttpStatus.NOT_FOUND)
        .contentType(MediaType.APPLICATION_JSON)
        .body(NOT_FOUND_BODY);
  }
}

File: profiles/kotlin/scaffold/src/main/kotlin/com/example/app/NotFoundAdvice.kt
package com.example.app

import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.NoHandlerFoundException

/**
 * Maps an unrouted request to the contract's hardened JSON 404 (Kotlin/Spring
 * profile).
 *
 * With `spring.web.resources.add-mappings=false` +
 * `spring.mvc.throw-exception-if-no-handler-found=true` (see
 * `application.properties`), an unknown GET path raises [NoHandlerFoundException]
 * instead of returning the Whitelabel page; this advice turns it into
 * `{"error":"not found"}` with `application/json`, so the body matches the go/python
 * references EXACTLY. Non-GET methods never reach here — the
 * [SecurityAndTelemetryFilter] short-circuits them to the same 404 before dispatch.
 *
 * The security headers + per-request telemetry are still applied: this handler runs
 * INSIDE the filter's wrap, so the filter's stamped headers and after-response
 * telemetry cover this 404 too.
 */
@RestControllerAdvice
class NotFoundAdvice {
    @ExceptionHandler(NoHandlerFoundException::class)
    fun handleNoHandler(): ResponseEntity<Map<String, String>> =
        ResponseEntity.status(HttpStatus.NOT_FOUND)
            .contentType(MediaType.APPLICATION_JSON)
            .body(mapOf("error" to "not found"))
}
